documentation/writing_tests.md

```diff
@@ -0,0 +1,1473 @@
+
+An example of a Karma file that can be used can be seen in the following path: <https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/karma.conf.js>
+
+You should pay special attention to the lines with the following content:
+
+- Ensure that the built browser output, made by rollup, is referenced as an element of the `files` array property. It should look like `dist-test/index.browser.js`, [as you can see in the template file](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/karma.conf.js#L31).
+- The `preprocessors` property object should specify what files will have test coverage, which should look like `"dist-test/index.browser.js": ["coverage"]`, and what files can be imported/required even though they're JSON files, which should look like `"recordings/browsers/**/*.json": ["json"]`.
+- Make sure that the timeouts are reasonably similar to the timeouts specified on your Mocha configuration. You will be placing the same number of milliseconds on the properties: `browserNoActivityTimeout`, and in `client: { mocha: { timeout: /* here */ } }`, and you can specify less milliseconds on `browserDisconnectTimeout`.
+
+Besides the default contents of that Karma configuration file, you can consider the following additions:
+
+- You might benefit from loading [`dotenv`](https://www.npmjs.com/package/dotenv), to specify your environment variables in a local file that shouldn't be tracked by Git. If you decide to use that extra dependency, make sure to load the `.env` file at the top-most of the file, right after any imports or requires. The line that loads the `.env` file might look like: `require("dotenv").config({ path: "../.env" });`.
+- In some computers, specially Linux boxes without a graphical environment, tests might not run unless you specifically load a headless Chrome, which should look like this:
+```js
+  browsers: ["ChromeHeadlessNoSandbox"],
+  customLaunchers: {
+    ChromeHeadlessNoSandbox: {
+        base: "ChromeHeadless",
+        flags: ["--no-sandbox"]
+    }
+  }
+```
+- Most of our tests focus on checking that our code preemptively works on browsers, even before our services support authentication methods compatible with browsers. For this reason, it's important to disable the browser's web security when necessary. This is done by sending the flag `--disable-web-security`. Combined with `ChromeHeadlessNoSandbox`, it should look like this:
+```js
+  browsers: ["ChromeHeadlessNoSandbox"],
+  customLaunchers: {
+    ChromeHeadlessNoSandbox: {
+        base: "ChromeHeadless",
+        flags: ["--no-sandbox", "--disable-web-security"]
+    }
+  }
+```
+- You might want to avoid logging from the browser, since [the recorder](#the-recorder) uses the logging output to create files. You can omit logs using the `browserConsoleLogOptions` property, which you can conditionally apply, as in the following snippet:
+```js
+  browserConsoleLogOptions: {
+    terminal: !process.env.TEST_MODE
```

Comment:
There is no real reason someone would want to use this.
This snippet doesn't have to be in this doc at all if not for the following reasoning. 

- This is tightly coupled with `jsonRecordingFilterFunction` function.
- `jsonRecordingFilterFunction` 
   - works only in `record` mode, no-op in other modes
   - captures all the console logs, prints the console logs straight to the terminal if they are unrelated to recordings, the ones related to recordings are saved into files as the browser recordings.
   - to make the above mechanism work and simultaneously not corrupt the logs, we disable the logging with `browserConsoleLogOptions.terminal` since `jsonRecordingFilterFunction` takes care of logging into the terminal.
- And since the filter-function only works in `record` mode, we disable `browserConsoleLogOptions.terminal` only in record mode. 

This is just to explain how things are working here. I'd suggest you make it `terminal: process.env.TEST_MODE !== "record"` and it lines up with the updated description from my previous suggestion.

Reply:
Thank you!